services/requester/cloud_kms_rotation_signer.go

```diff
@@ -0,0 +1,104 @@
+package requester
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/onflow/flow-go-sdk/crypto"
+	"github.com/onflow/flow-go-sdk/crypto/cloudkms"
+	"github.com/rs/zerolog"
+)
+
+var _ crypto.Signer = &CloudKMSKeyRotationSigner{}
+
+// CloudKMSKeyRotationSigner is a crypto signer that contains a pool of
+// `crypto.Signer`[1] objects, each of which is tied to a Cloud KMS
+// asymmetric signing key.
+// It keeps track of the signer/key combination that should be used for
+// the next incoming signing request. This allows for faster submission
+// of transactions to the network, due to a sequence number not being
+// reused between different keys used.
+// It also contains logic to queue up signature requests and in case
+// there are more transactions pending than the keys in the pool it
+// will wait for transactions to get executed so the new sequence key
+// can be obtained.
+// The signer is concurrency-safe.
+//
+// [1](https://github.com/onflow/flow-go-sdk/blob/master/crypto/cloudkms/signer.go#L37)
+type CloudKMSKeyRotationSigner struct {
+	mux        sync.RWMutex
+	kmsSigners []*cloudkms.Signer
+	index      int
+	signersLen int
+	logger     zerolog.Logger
+}
+
+// NewSignerForKeys returns a new CloudKMSKeyRotationSigner, for the
+// given slice of Cloud KMS keys.
+func NewSignerForKeys(
+	ctx context.Context,
+	client *cloudkms.Client,
+	keys []cloudkms.Key,
+	logger zerolog.Logger,
+) (*CloudKMSKeyRotationSigner, error) {
+	if len(keys) == 0 {
+		return nil, fmt.Errorf("no asymmetric signing keys provided")
+	}
+
+	kmsSigners := make([]*cloudkms.Signer, len(keys))
+	for i, key := range keys {
+		kmsSigner, err := client.SignerForKey(ctx, key)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"could not create signer for key with ID: %s: %w",
+				key.KeyID,
+				err,
+			)
+		}
+		kmsSigners[i] = kmsSigner
+	}
+
+	logger = logger.With().Str("component", "cloud_kms_signer").Logger()
+
+	return &CloudKMSKeyRotationSigner{
+		kmsSigners: kmsSigners,
+		signersLen: len(kmsSigners),
+		logger:     logger,
+	}, nil
+}
+
+// Sign signs the message and then rotates to the next key.
+// Note: if you want to get the public key pair, you should first call
+// PublicKey and then Sign.
+func (s *CloudKMSKeyRotationSigner) Sign(message []byte) ([]byte, error) {
+	defer s.timeTrack(time.Now())
```

Comment:
@sideninja I have added logging for the duration of each message signing request, e.g.:

```json
{
  "level": "info",
  "component": "cloud_kms_signer",
  "duration": 118,
  "time": "2024-06-10T07:00:05Z",
  "message": "messaged was signed"
}
```
The `duration` being in milliseconds. Do we need to add more info? Also, I am taking in account the lock-acquisition as well. Should we maybe track only the actual request to Cloud KMS instead?

Reply:
good idea to keep track of time, but wouldn't extract it out, look at the comment bellow. I also wouldn't make it `info` level, this is more of a `debug` level.

Reply:
Updated in https://github.com/onflow/flow-evm-gateway/pull/282/commits/381b26ca17813222c86af0352f6b32f8f6e9391a, to log as `debug` level.